### PR TITLE
Cleanup of namespacet::lookup vs symbol_tablet::lookup

### PR DESCRIPTION
**Pull request by tautschnig**
This commit series was sparked by the great observation of @reuk and the proposed fix in #711. These patches take a different approach of removing the code. One of #711 or this one will have to be closed. I'm putting this one up for discussion.

### CODE REVIEW

**Comment by peterschrammel** [comment]
I am in favour of this one, but  it most also be applied to cegis first. https://travis-ci.org/diffblue/cbmc/jobs/216054189#L1671

**Comment by tautschnig** [comment]
> However, I'm against this change. As I understand it, symbol_tablet exposes the API to maintain a map of symbols (add, remove, has_symbol, clear). Surely being able to also find a symbol should remain part of that interface?

Yes, this seems very reasonable. But then I'd still suggest to not overload `lookup` in a modifying fashion. How about keeping a `const lookup` method, actually _use_ that from `namespacet`, and add a method for obtaining a writable `symbolt`?

**Comment by tautschnig** [comment]
@pkesseli Where can I find the cegis repository so that I could send you a pull request?

**Comment by pkesseli** [comment]
@tautschnig Please find it here: https://github.com/diffblue/cegis

**Comment by tautschnig** [comment]
> @tautschnig Please find it here: https://github.com/diffblue/cegis

That's 404 for me, i.e., it isn't a public repo.

**Comment by tautschnig** [comment]
I believe this is now at the state that's agreed.

**Comment by peterschrammel** [comment]
@pkesseli, is this ok with you now?

**Comment by pkesseli** [comment]
The only thing I would appreciate is a const overload of `symbolt &get(const irep_idt &identifier);`.

**Comment by pkesseli** [comment]
Also, I assume we rename `lookup` in `symbol_tablet` to `get` is to differentiate it from an actual namspace lookup? Should we make it an `operator[](const irep_idt &)` then?

**Comment by tautschnig** [comment]
I hope to have addressed the problems; @pkesseli does the API look sane now?

File: src/musketeer/propagate_const_function_pointers.cpp [hunk 1]
```diff
@@ -374,7 +376,7 @@ void const_function_pointer_propagationt::arg_stackt::add_args(
       const symbol_exprt &arg_symbol_expr=to_symbol_expr(arg_expr);
 
       // const symbolt &arg_symbol=
-        // cfpp.symbol_table.lookup(arg_symbol_expr.get_identifier());
+        // cfpp_ns.lookup(arg_symbol_expr.get_identifier());
```
Comment (peterschrammel) [inline src/musketeer/propagate_const_function_pointers.cpp]: Unrelated: use #if 0 or delete

File: src/goto-programs/remove_exceptions.cpp [hunk 2]
```diff
@@ -40,6 +41,7 @@ class remove_exceptionst
 
 protected:
   symbol_tablet &symbol_table;
+  namespacet ns;
```
Comment (peterschrammel) [inline src/goto-programs/remove_exceptions.cpp]: Having both is a bit ugly...

**Comment by peterschrammel** [comment]
I agree with @pkesseli to use `operator[](const irep_idt &)` instead of `get`

**Comment by tautschnig** [comment]
> I agree with @pkesseli to use `operator[](const irep_idt &)` instead of `get`

Shouldn't that be `at(const irep_idt&)` instead of `operator[]` as it doesn't have any of the `[]` semantics?

**Comment by tautschnig** [comment]
@peterschrammel @pkesseli My apologies for leaving this for so long - would you mind taking another look?

**Comment by tautschnig** [comment]
Closing as symbol tables have gotten a completely different API ever since.

**Pull request closed without merging**